src/Lean/Meta/Tactic/Apply.lean

```diff
@@ -230,4 +230,22 @@ def _root_.Lean.MVarId.exfalso (mvarId : MVarId) : MetaM MVarId :=
     mvarId.assign (mkApp2 (mkConst ``False.elim [u]) target mvarIdNew)
     return mvarIdNew.mvarId!
 
+/--
+Apply the `n`-th constructor of the target type,
+checking that it is an inductive type,
+and that there are the expected number of constructors.
+-/
+def _root_.Lean.MVarId.nthConstructor
+    (name : Name) (idx : Nat) (expected? : Option Nat := none) (goal : MVarId) :
+    MetaM (List MVarId) := do
+  goal.withContext do
+    goal.checkNotAssigned name
+    matchConstInduct (← goal.getType').getAppFn
+      (fun _ => throwTacticEx name goal "target is not an inductive datatype")
+      fun ival us => do
+        if let some e := expected? then unless ival.ctors.length == e do
+          throwTacticEx name goal
+            s!"{name} tactic works for inductive types with exactly {e} constructors"
+        goal.apply <| mkConst ival.ctors[idx]! us
```

Comment:
If this is called without `expected?`, it will panic. The only uses here are `left` and `right`, which do set `expected`.  Is it worth making this function safer by requiring `expected?` and maybe even setting `idx : Fin expected`? Or else at least check `idx < ival.ctrs.size` here?

Reply:
Having `expected?` be optional allows this to be used for tactics that work for more general types. But I think we should not generalize to cases that we may someday want, as it makes the code harder to use if we don't actually want to support them. In fact, I think this could even be simplified to a version that supports just 2-constructor datatypes, unless we have further plans to upstream more general tactics in this family.

Reply:
I think just an informative error unless `idx < ival.ctrs.size`. Want to push the change? :-)

Reply:
Can do after breakfast

Reply:
```suggestion
        unless (idx < ival.ctors.length) do
          throwError m!"nthConstructor: idx {idx} out of bounds, only {ival.ctors.length} constructors"
        goal.apply <| mkConst ival.ctors[idx]! us
```